### DISCUSSION TITLE
listens_for(Engine, "before_cursor_execute") does not work for Core, only ORM

### DISCUSSION THREAD

**Discussion by fvant** [sqlalchemy/sqlalchemy#10736] [Usage Questions] [answered]
The recipe for query profiling in the FAQ, using before_cursor_execute and after_cursor_execute @event.listens_for works for ORM queries but i did not get them to work for Core queries.

As i used Engine, i expected both to work.

Tried pymysql and mysqldb drivers with same result.
sqlalchemy 2.2.0
python 3.11 on OSX

**Comment by zzzeek** [comment]
hi -

taking the example at https://docs.sqlalchemy.org/en/20/faq/performance.html#query-profiling as written, and adding a short sequence using an engine is below:


```py
from sqlalchemy import event
from sqlalchemy.engine import Engine
import time
import logging

logging.basicConfig()
logger = logging.getLogger("myapp.sqltime")
logger.setLevel(logging.DEBUG)


@event.listens_for(Engine, "before_cursor_execute")
def before_cursor_execute(conn, cursor, statement, parameters, context, executemany):
    conn.info.setdefault("query_start_time", []).append(time.time())
    logger.debug("Start Query: %s", statement)


@event.listens_for(Engine, "after_cursor_execute")
def after_cursor_execute(conn, cursor, statement, parameters, context, executemany):
    total = time.time() - conn.info["query_start_time"].pop(-1)
    logger.debug("Query Complete!")
    logger.debug("Total Time: %f", total)


from sqlalchemy import create_engine

e = create_engine("mysql://scott:tiger@localhost/test")

with e.connect() as conn:
    result = conn.exec_driver_sql("select 1")
    print(result.all())
```

output of this program shows:

```
DEBUG:myapp.sqltime:Start Query: select 1
DEBUG:myapp.sqltime:Query Complete!
DEBUG:myapp.sqltime:Total Time: 0.000217
[(1,)]

```


try that program out and see what might be different about your own application.

**Comment by fvant** [comment]
apologies, i did forget to detail my use case.

I do use cursors like so:
` with conn.cursor(MySQLdb.cursors.Dictcursor) as cursor: `

and to get cursors, I use
``` 
session = sessionmaker(eng)
conn = session().connection().connection
```

**Reply by zzzeek** [reply, marked as answer]
SQLAlchemy's events dont instrument on top of the underlying cursor of the DBAPI, so to profile queries emitted via direct cursor access you would need to write your own start/stop endpoints.

**Reply by CaselIT** [reply]
you may look into `exec_driver_sql` to pass a sql string unchanged to the driver. Using that method will use the sqlalchemy event system
